frontends/PyCDE/src/pycde/esi.py

```diff
@@ -929,6 +929,73 @@ def _op(sym_name: ir.StringAttr):
 Telemetry = _Telemetry()
 
 
+class TelemetryMMIO(ServiceImplementation):
+  """An ESI service implementation which provides telemetry data through an MMIO
+  region. Each client request is assigned a register in the MMIO space. When a
+  read request is received for the assigned address, it gets routed to the
+  assigned client. When a write request is received, it is discarded. The
+  assignment table is stored in the manifest."""
+
+  clk = Clock()
+  rst = Reset()
+
+  @generator
+  def generate(ports, bundles: _ServiceGeneratorBundles) -> bool:
+    if len(bundles.to_client_reqs) == 0:
+      # No clients to connect to, so we don't need to do anything.
+      return True
+
+    mmio_cmd = MMIO.read_write(AppID("__telemetry_mmio"))
+    # Assign each telemetry client a register offset in MMIO space.
+
+    offset = 0
+    table: Dict[int, AssignableSignal] = {}
+    for bundle in bundles.to_client_reqs:
+      # Only support 'report' port for telemetry.
+      if bundle.port == 'report':
```

Comment:
Would be nice that these things are centralized/documented somewhere, so that we don't have these stray magic strings.